docs/source/developing-volttron/developing-agents/using-concurrency-in-agents.rst

```diff
@@ -0,0 +1,130 @@
+
+===================================
+Using Concurrency in Agents
+===================================
+
+When would you want to use concurrency in your agent? You might want to use concurrency if you want to improve performance.
+For example, your agent performs an I/O-bound task such as scraping a website or talking to a slow device (e.g. hard drive or printer)
+and you notice it takes a significant amount to time to complete such a task.
+Or you might have to use concurrency because your agent uses a dependency that has concurrency--which forces your agent
+code to be concurrent.
+
+Regardless of your motivations, using concurrency to develop your Agent is not a trivial task.
+To help you implement concurrency using Python within the Agent development framework, we have provided the following
+example from the OpenADRVenAgent and accompanying explanations. Please note that the example uses one type of
+concurrency, `cooperative multitasking <https://en.wikipedia.org/wiki/Cooperative_multitasking>`_. If you need an
+example of how to implement other types of concurrency, please see `this tutorial <https://realpython.com/async-io-python/>`_.
+
+Before we dive into the example, we establish the following concepts:
+
+* An Event Loop is a software design pattern that handles events concurrently; it waits for and dispatches multiple events concurrently and gives the illusion of executing the events in "parallel". In Python, the `Event Loop <https://docs.python.org/3.10/library/asyncio-eventloop.html#event-loop>`_ contains a list of Tasks which controls when and how those Tasks are executed.
+
+* A `Task <https://docs.python.org/3/library/asyncio-task.html#task-object>`_ is an object that runs a `coroutine <https://docs.python.org/3/library/asyncio-task.html#coroutine>`_ (i.e. asynchronous function). In Python, coroutines are written using the 'async' keyword.
+
+* Asyncio is a built-in Python module that allows the developer to write concurrent code.
+
+* Gevent is a "coroutine-based Python networking library that uses greenlet to provide a high-level synchronous API on top of the libev or libuv event loop" (see `Gevent docs <http://www.gevent.org/>`_. It is an alternative to Asyncio.
+
+The OpenADRVenAgent demonstrates how to write concurrent code in an agent; it uses Asyncio and Gevent.
+It is not required to use both modules to write concurrency in your agent. The purpose of this example is two-fold: 1) to
+show an example of using concurrency in an agent and 2) to show how Gevent can be used with Asyncio.
+
+
+The OpenADRVenAgent has a coroutine called 'handle_event' which will be one of the Tasks that will be executed during the Event Loop. An example of the
+coroutine is shown below:
+
+.. code-block:: python
+
+    async def handle_event(self, event):
+
+        # do things
+
+
+The OpenADRVenAgent has a VEN client and this coroutine is added to that client. To create the Event Loop, the OpenADRVenAgent has a helper method
```

Comment:
This is out of context, as the loop is actually running inside a gevent spawned coroutine.